tokenscan/scanner.go

```diff
@@ -0,0 +1,221 @@
+	errors         prometheus.Counter
+}
+
+// Opt is an option func
+type Opt func(*Scanner)
+
+// WithParallelism configures the scan parallelism
+func WithParallelism(parallelism int) Opt {
+	return func(s *Scanner) {
+		s.parallelism = parallelism
+	}
+}
+
+// New creates a new scylla token range scanner
+func New(sessionBuilder SessionBuilder, opts ...Opt) *Scanner {
+	s := &Scanner{
+		sessionBuilder: sessionBuilder,
+
+		parallelism: defaultParallelism,
+		results: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "scyllascanner_results_total",
+			Help: "Number of items returned by the scylla scanner",
+		}),
+		errors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "scyllascaller_errors_total",
+			Help: "Number of errors encountered while scanning",
+		}),
+	}
+	for _, o := range opts {
+		o(s)
+	}
+
+	s.tokenRanges = getShuffledTokenRanges(s.parallelism)
+	return s
+}
+
+// Describe implements the prometheus.Collector interface
+func (s *Scanner) Describe(ch chan<- *prometheus.Desc) {
+	s.results.Describe(ch)
+	s.errors.Describe(ch)
+}
+
+// Collect implements the prometheus.Collector interface
+func (s *Scanner) Collect(ch chan<- prometheus.Metric) {
+	s.results.Collect(ch)
+	s.errors.Collect(ch)
+}
+
+// Scan performs a token range scan with the provided select builder
+// The scanned items are passed to the ApplyFunc. If the ApplyFunc returns ErrAbort the scanning process
+// is aborted. All other ApplyFunc errors are just logged
+func (s *Scanner) Scan(
+	ctx context.Context,
+	selectBuilder *qb.SelectBuilder,
+	partitionKeyColumns []string,
+	applyFunc ApplyFunc,
+) error {
+	// send all available token ranges to a channel
+	rangesChannel := make(chan *tokenRange, len(s.tokenRanges))
+	for i := range s.tokenRanges {
+		rangesChannel <- s.tokenRanges[i]
+	}
+	close(rangesChannel)
+
+	// create a cancelable context
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	var sessionCreationWaitGroup sync.WaitGroup
+	errGroup := errgroup.Group{}
+	sessionCreationWaitGroup.Add(s.parallelism)
+
+	iterRows := make(chan map[string]interface{}) // TODO: channel cap?
+	defer close(iterRows)
+
+	go func() {
+		for row := range iterRows {
+			if abort := applyFunc(ctx, row); abort {
+				cancel()
+				return
+			}
+		}
+	}()
+
+	for i := 0; i < s.parallelism; i++ {
+		errGroup.Go(func() error {
+			session, err := s.sessionBuilder.CreateSession()
+			if err != nil {
+				cancel()
+				return err
+			}
+			defer session.Close()
+
+			// synchronize start of token range scans, i.e. wait for all sessions to be created
+			sessionCreationWaitGroup.Done()
+			// Read ranges from the channel, until we have completed accessing all ranges
+			for r := range rangesChannel {
+				stmt, names := tokenSelect(selectBuilder, partitionKeyColumns...).ToCql()
+				q := gocqlx.Query(session.Query(stmt, names).WithContext(ctx), names).BindMap(qb.M{
+					"lower": r.startRange,
+					"upper": r.endRange,
+				})
+				iter := q.Iter()
```

Comment:
Compare that with https://github.com/scylladb/gocqlx/blob/a72115182914eced6612440463a3ab40c727c5b4/example_test.go#L586